types/src/v0/v0_1/l1.rs

```diff
@@ -33,12 +38,88 @@ pub struct L1Snapshot {
+    /// Delay when retrying failed L1 queries.
+    #[clap(
+        long,
+        env = "ESPRESSO_SEQUENCER_L1_RETRY_DELAY",
+        default_value = "1s",
+        value_parser = parse_duration,
+    )]
+    pub l1_retry_delay: Duration,
+
+    /// Request rate when polling L1.
+    #[clap(
+        long,
+        env = "ESPRESSO_SEQUENCER_L1_POLLING_INTERVAL",
+        default_value = "7s",
+        value_parser = parse_duration,
+    )]
+    pub l1_polling_interval: Duration,
+
+    /// Maximum number of L1 blocks to keep in cache at once.
+    #[clap(long, env = "ESPRESSO_SEQUENCER_L1_BLOCKS_CACHE_SIZE", default_value = "100")]
+    pub l1_blocks_cache_size: NonZeroUsize,
+
+    /// Number of L1 events to buffer before discarding.
+    #[clap(long, env = "ESPRESSO_SEQUENCER_L1_EVENTS_CHANNEL_CAPACITY", default_value = "100")]
+    pub l1_events_channel_capacity: usize,
+
+    /// Maximum number of L1 blocks that can be scanned for events in a single query.
+    #[clap(
+        long,
+        env = "ESPRESSO_SEQUENCER_L1_EVENTS_MAX_BLOCK_RANGE",
+        default_value = "10000"
+    )]
+    pub l1_events_max_block_range: u64,
+}
+
 #[derive(Clone, Debug)]
-/// An Http Provider and configuration to interact with the L1.
+/// An Ethereum provider and configuration to interact with the L1.
+///
+/// This client runs asynchronously, updating an in-memory snapshot of the relevant L1 information
+/// each time a new L1 block is published. The main advantage of this is that we can update the L1
+/// state at the pace of the L1, instead of the much faster pace of HotShot consensus.This makes it
+/// easy to use a subscription instead of polling for new blocks, vastly reducing the number of L1
+/// RPC calls we make.
 pub struct L1Client {
-    pub retry_delay: Duration,
+    pub(crate) retry_delay: Duration,
     /// `Provider` from `ethers-provider`.
-    pub provider: Arc<Provider<Http>>,
+    pub(crate) provider: Arc<Provider<RpcClient>>,
     /// Maximum number of L1 blocks that can be scanned for events in a single query.
-    pub events_max_block_range: u64,
+    pub(crate) events_max_block_range: u64,
+    /// Shared state updated by an asynchronous task which polls the L1.
+    pub(crate) state: Arc<Mutex<L1State>>,
+    /// Channel used by the async update task to send events to clients.
+    pub(crate) sender: Sender<L1Event>,
+    /// Receiver for events from the async update task.
+    pub(crate) receiver: InactiveReceiver<L1Event>,
+    /// Async task which updates the shared state.
+    pub(crate) update_task: Arc<L1UpdateTask>,
+}
+
+/// An Ethereum RPC client over HTTP or WebSockets.
+#[derive(Clone, Debug)]
+pub(crate) enum RpcClient {
+    Http(Http),
+    Ws(Ws),
+}
+
+/// In-memory view of the L1 state, updated asynchronously.
+#[derive(Debug)]
+pub(crate) struct L1State {
+    pub(crate) snapshot: L1Snapshot,
+    pub(crate) finalized: LruCache<u64, L1BlockInfo>,
+    pub(crate) finalized_by_timestamp: BTreeMap<U256, u64>,
```

Comment:
Similar to my last comment, IIUC since is only used on startup and in-memory only I think it might be simpler if we didn't have it at all or at least didn't keep a cache for it.

Reply:
Good call, removing this cache makes things considerably simpler at very little cost